File: backend/app/models/biometric_consent.py
"""
Biometric Consent Model.

COPPA-compliant biometric data consent tracking for Zeh Ani features.
Tracks granular consent for mesh generation, voice-to-voice, and
latent feature extraction separately per child profile.
"""

from datetime import datetime, timezone
from enum import Enum
from typing import Optional

from beanie import Document, Indexed
from pydantic import Field


class BiometricConsentType(str, Enum):
    """Types of biometric data processing requiring separate consent."""

    MESH_GENERATION = "mesh_generation"
    VOICE_V2V = "voice_v2v"
    LATENT_FEATURES = "latent_features"


class BiometricConsent(Document):
    """
    Granular biometric consent record per child profile.

    Each consent type (mesh, voice, latent) is tracked independently
    to allow parents fine-grained control over biometric data usage.
    Family PIN verification is required for all consent operations.
    """

    user_id: Indexed(str)
    profile_id: str
    consent_type: BiometricConsentType

    # Consent metadata
    granted_at: datetime = Field(
        default_factory=lambda: datetime.now(timezone.utc)
    )
    granted_by_user_id: str
    family_pin_verified: bool = False

    # Data residency preferences
    on_device_only: bool = Field(
        default=True,
        description="If True, biometric data stays on-device only",
    )
    latent_features_cloud: bool = Field(
        default=False,
        description="If True, latent features may be stored in cloud",
    )

    # Revocation
    revoked_at: Optional[datetime] = None

    class Settings:
        name = "biometric_consents"
        indexes = [
            [("user_id", 1), ("profile_id", 1), ("consent_type", 1)],
        ]

    @property
    def is_active(self) -> bool:
        return (
            self.family_pin_verified
            and self.revoked_at is None
        )

File: android-app/app/src/main/java/tv/bayit/plus/data/model/zehani/BiometricConsent.kt
package tv.bayit.plus.data.model.zehani

import com.google.gson.annotations.SerializedName

data class BiometricConsentStatus(
    @SerializedName("profile_id")
    val profileId: String,

    @SerializedName("consents")
    val consents: List<ConsentEntry>
)

data class ConsentEntry(
    @SerializedName("consent_type")
    val consentType: BiometricConsentType,

    @SerializedName("active")
    val active: Boolean
)

enum class BiometricConsentType {
    @SerializedName("mesh_generation") MESH_GENERATION,
    @SerializedName("voice_v2v") VOICE_V2V,
    @SerializedName("latent_features") LATENT_FEATURES
}

data class BiometricConsentRequest(
    @SerializedName("profile_id")
    val profileId: String,

    @SerializedName("consent_type")
    val consentType: String,

    @SerializedName("pin")
    val pin: String
)

data class BiometricConsentResponse(
    @SerializedName("profile_id")
    val profileId: String,

    @SerializedName("consent_type")
    val consentType: String,

    @SerializedName("active")
    val active: Boolean,

    @SerializedName("granted_at")
    val grantedAt: String
)
